plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs

```diff
@@ -258,28 +334,56 @@ For a (self-)recursive datatype we have to change three things:
 - Destructors must include an unwrap.
 -}
 
--- See note [Scott encoding of datatypes]
+-- See note [Encoding of datatypes]
 -- | Make the "Scott-encoded" type for a 'Datatype', with type variables free.
+-- This is exactly the type of an eliminator function for the datatype.
+--
 -- @mkScottTy Maybe = forall out_Maybe. out_Maybe -> (a -> out_Maybe) -> out_Maybe@
-mkScottTy :: ann -> Datatype TyName Name uni ann -> TyName -> Type TyName uni ann
-mkScottTy ann (Datatype _ _ _ _ constrs) res =
+mkScottTy :: MonadQuote m => ann -> Datatype TyName Name uni ann -> m (Type TyName uni ann)
+mkScottTy ann d@(Datatype _ _ _ _ constrs) = do
+   res <- resultTypeName d
+   -- FIXME: normalize datacons' types also here
+   let caseTys = fmap (constructorCaseType ann (TyVar ann res)) constrs
+   pure $
       -- forall res.
       TyForall ann res (Type ann) $
-          -- c_1 -> ... -> c_n -> res
-          PIR.mkIterTyFun ann caseTys (TyVar ann res)
-  where
-     -- FIXME: normalize datacons' types also here
-    caseTys = fmap (constructorCaseType (TyVar ann res)) constrs
-
--- | Make the real PLC type corresponding to a 'Datatype' with the given pattern functor.
--- @
---     mkDatatypeType List <pattern functor of List>
---         = fix list . <pattern functor of List>
---         = fix list . \(a :: *) -> forall (r :: *) . r -> (a -> List a -> r) -> r
--- @
-mkDatatypeType :: forall m uni fun a. MonadQuote m => Recursivity -> PIRType uni a -> Datatype TyName Name uni (Provenance a) -> m (PLCRecType uni fun a)
-mkDatatypeType r pf (Datatype ann tn tvs _ _) = case r of
-    NonRec -> PlainType <$> (PLC.mkIterTyLam <$> pure tvs <*> pure pf)
+      -- c_1 -> .. -> c_n -> res
+      PIR.mkIterTyFun ann caseTys (TyVar ann res)
+
+mkDatatypeSOPTy :: ann -> Datatype TyName Name uni ann -> Type TyName uni ann
+mkDatatypeSOPTy ann (Datatype _ _ _ _ constrs) = TySOP ann (fmap constructorArgTypes constrs)
+
+-- | Make the "pattern functor" of a 'Datatype'. This is just the normal type, but with the
+-- type variable for the type itself free and its type variables free.
```

Comment:
Not quite! The pattern functor for, say, `list` is

```
listF = \(list :: * -> *) (a :: *) -> all (r :: *). r -> (a -> list a -> r) -> r
```

with `list` defined in terms of it as

```
list = \(a :: *) -> ifix listF a
```

There's no free variables in there anywhere. If we strip lambdas off of `listF`, we'll get "the body of the pattern functor". Well, according to the terminology I've been using anyway, maybe I've been butchering it, not sure.